### PR TITLE
refactor: Finalize message handling refactoring

### PR DESCRIPTION
**Pull request by matz3**
- Use new addMessage signature in all reporters
- Reduce duplicate code for addMessage by moving logic into LinterContext
- Only produce actual messages at the end of linting (Might slightly reduce peak memory usage)

### CODE REVIEW

**Review by RandomByte — COMMENTED** [review]
Looks like some solid improvements 👍 
Maybe @d3xter666 wants to have a look too, otherwise fine with me.

File: test/lib/linter/snapshots/linter.ts.md [hunk 1]
```diff
@@ -446,41 +445,39 @@ Generated by [AVA](https://avajs.dev).
...
+            message: 'Use of deprecated property \'synchronizationMode\' of class \'sap.ui.model.odata.v4.ODataModel\'',
+            messageDetails: 'As of Version 1.110.0, this parameter is obsolete.',
             ruleId: 'ui5-linter-no-deprecated-api',
             severity: 2,
           },
           {
             column: 21,
-            fatal: undefined,
             line: 78,
-            message: 'Use of deprecated property \'sap.ui5/models/odata-v4-via-dataSource/settings/synchronizationMode\' of sap.ui.model.odata.v4.ODataModel',
```
Comment (d3xter666) [inline test/lib/linter/snapshots/linter.ts.md:475]: It seems that the path to the dataSource got lost and only the property is left.
Was this intentional?
I'm fine with it as the positiinalInfo (line/column) is correct

**Review by d3xter666 — APPROVED** [review]
Overall LGTM
I have some questions as I might be missing some of the details

File: src/linter/LinterContext.ts [hunk 2]
```diff
@@ -177,13 +190,50 @@ export default class LinterContext {
...
+			if (messageDetails) {
+				message.messageDetails = resolveLinks(messageDetails);
+			}
+		}
+
+		if ("fatal" in messageInfo && typeof messageInfo.fatal === "boolean") {
+			message.fatal = messageInfo.fatal;
+		}
+
+		if (message.fatal && message.severity !== LintMessageSeverity.Error) {
```
[the remark this answers is not in the document]
Comment (matz3) [reply src/linter/LinterContext.ts:221]: No, you're not. We previously had this check and I just kept it. We could of course replace it with a unit tests that checks this constraint for all messages, but the runtime overhead for the check looks very small to me.

File: src/linter/html/HtmlReporter.ts [hunk 3]
```diff
@@ -19,24 +16,27 @@ export default class HtmlReporter {
...
 
-	addMessage({node, message, messageDetails, severity, ruleId, fatal = undefined}: ReporterMessage) {
-		if (fatal && severity !== LintMessageSeverity.Error) {
-			throw new Error(`Reports flagged as "fatal" must be of severity "Error"`);
+	addMessage<M extends MESSAGE>(id: M, args: MessageArgs[M], node: SaxTag): void;
+	addMessage<M extends MESSAGE>(id: M, node: SaxTag): void;
+	addMessage<M extends MESSAGE>(
+		id: M, argsOrNode?: MessageArgs[M] | SaxTag, node?: SaxTag
+	) {
+		if (!argsOrNode) {
```
[the remark this answers is not in the document]
Comment (matz3) [reply src/linter/html/HtmlReporter.ts:24]: To me this all seems required to distinguish between the signatures at runtime. Feel free to improve this part. I couldn't manage to make it better. The only alternative would be to not have optional arguments at all, but then it was hard for me to come up with a type for the messages without args.

File: test/lib/linter/snapshots/linter.ts.md (hunk 1 above)
Comment (matz3) [reply test/lib/linter/snapshots/linter.ts.md:475]: Good catch. The message was better before.

File: src/linter/LinterContext.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (d3xter666) [reply src/linter/LinterContext.ts:221]: Thanks!

**Review by d3xter666 — APPROVED** [review]
LGTM

**Pull request merged**